File: ITEM 20-29/ITEM26 OpenCV_SelectiveSearch/OpenCV_SelectiveSearch.cpp
#include "pch.h"
#include <opencv2/opencv.hpp>
#include <opencv2/ximgproc.hpp>
#include <iostream>
#include <ctime>

using namespace cv;
using namespace cv::ximgproc::segmentation;

int main()
{
	// speed-up using multithreads 使用多线程
	//开启CPU的硬件指令优化功能
	setUseOptimized(true);
	setNumThreads(4);

	// read image 读图
	Mat im = imread("./image/dogs.jpg");
	if (im.empty())
	{
		return 0;
	}
	// resize image 图像大小重置
	int newHeight = 200;
	int newWidth = im.cols*newHeight / im.rows;
	resize(im, im, Size(newWidth, newHeight));

	// create Selective Search Segmentation Object using default parameters 默认参数生成选择性搜索类
	Ptr<SelectiveSearchSegmentation> ss = createSelectiveSearchSegmentation();
	// set input image on which we will run segmentation 要进行分割的图像
	ss->setBaseImage(im);

	// Switch to fast but low recall Selective Search method 快速搜索(速度快，召回率低)
	//ss->switchToSelectiveSearchFast();
	//精准搜索(速度慢，召回率高)
	ss->switchToSelectiveSearchQuality();

	// run selective search segmentation on input image 保存搜索到的框，按可能性从高到低排名
	std::vector<Rect> rects;
	ss->process(rects);
	std::cout << "Total Number of Region Proposals: " << rects.size() << std::endl;

	// number of region proposals to show 在图像中保存多少框
	int numShowRects = 100;

	while (1)
	{
		// create a copy of original image 做一份图像图像拷贝
		Mat imOut = im.clone();

		// itereate over all the region proposals 画框前numShowRects个
		for (int i = 0; i < numShowRects; i++)
		{
			rectangle(imOut, rects[i], Scalar(0, 255, 0));
		}

		// show output
		imshow("Output", imOut);

		waitKey(0);
	}
	return 0;
}

File: ITEM 20-29/ITEM26 OpenCV_SelectiveSearch/OpenCV_SelectiveSearch.py

import sys
import cv2

if __name__ == '__main__':

    # speed-up using multithreads
    cv2.setUseOptimized(True)
    cv2.setNumThreads(4);

    # read image
    im = cv2.imread('image/dogs.jpg')
    # resize image
    newHeight = 200
    newWidth = int(im.shape[1]*200/im.shape[0])
    im = cv2.resize(im, (newWidth, newHeight))    

    # create Selective Search Segmentation Object using default parameters
    ss = cv2.ximgproc.segmentation.createSelectiveSearchSegmentation()

    # set input image on which we will run segmentation
    ss.setBaseImage(im)


    #ss.switchToSelectiveSearchFast()

    ss.switchToSelectiveSearchQuality()


    # run selective search segmentation on input image
    rects = ss.process()
    print('Total Number of Region Proposals: {}'.format(len(rects)))
    
    # number of region proposals to show
    numShowRects = 100
    # increment to increase/decrease total number
    # of reason proposals to be shown
    increment = 50

    while True:
        # create a copy of original image
        imOut = im.copy()

        # itereate over all the region proposals
        for i, rect in enumerate(rects):
            # draw rectangle for region proposal till numShowRects
            if (i < numShowRects):
                x, y, w, h = rect
                cv2.rectangle(imOut, (x, y), (x+w, y+h), (0, 255, 0), 1, cv2.LINE_AA)
            else:
                break

        # show output
        cv2.imshow("Output", imOut);

        # record key press
        k = cv2.waitKey(0) & 0xFF

        # m is pressed
        if k == 109:
            # increase total number of rectangles to show by increment
            numShowRects += increment
        # l is pressed
        elif k == 108 and numShowRects > increment:
            # decrease total number of rectangles to show by increment
            numShowRects -= increment
        # q is pressed
        elif k == 113:
            break
    # close image show window
    cv2.destroyAllWindows()
